Fix Rule Evaluation Logic for Handling Multiple Rules of the Same Type

Vyom-Yadav opened the pull request:
Fixes #1609
Fixes #2119 

---

Major Changes and Implications:

- A new field, `name` is introduced to identify rules uniquely. The field would be optional in certain cases.
- The name is generated for migrating currently running databases in case of collisions using `rule_type`+`uuid`. 
- The rules are reconciled by the subsequent `InternalProfileInitEventTopic`, and `ExecuteEntityEventTopic` events. Profiles/Repositories not receiving these types of events **won't be reconciled** to update them to the latest state.
- Reconciliation can potentially open new PRs/Security Advisories, etc., as the duplicate rule's row (entry) was overridden by the other rule of the same type.
- Documentation specifies when `name` is optional and when not.

---

Note to Reviewers: Only the last two commits are under consideration. 3rd last commit is included for correct local testing.

eleftherias reviewed (changes requested):
Thank for the PR @Vyom-Yadav! This is looking great!

I have a small comment inline and a general question about the reconciliation.

When trying this out locally, I had an older profile entry in my DB with 2 `dependabot_configured` rules. 
After applying the migration and updating the server, I see that the JSON `contextual_rules` in `entity_profiles` is updated with the generated names for both dependabot rules, which is great.
I also see that I have 1 rule with the type `dependabot_configured` in `rule_evaluations` with the correctly generated rule name. 
However, the 2nd `dependabot_configured` rule is not in my `rule_evaluations`.

Is it by design that we only keep the one previously saved rule? Or should we create the second rule, even though there was no entry for it in `rule_evaluations` previously?

Vyom-Yadav commented:
>However, the 2nd dependabot_configured rule is not in my rule_evaluations.

`rule_evaluations` are created and upserted by the engine. The problem is that rules for entities are evaluated again only when we receive certain events, like I mentioned in the desc:

> The rules are reconciled by the subsequent `InternalProfileInitEventTopic`, and `ExecuteEntityEventTopic` events. Profiles/Repositories not receiving these types of events **won't be reconciled** to update them to the latest state.

---

>Is it by design that we only keep the one previously saved rule? Or should we create the second rule, even though there was no entry for it in rule_evaluations previously?

It is by our event-driven architecture. But I have mixed feelings about it. If we reconcile continuously and not based on events (internal, webhooks, etc.), we will consume a lot of resources as the reconciliation would continuously evaluate rules for entities. On the other hand, if we don't, then situations like these arise when the DB is not in the correct state after an update.  

The bottom line is with the current architecture and the way the engine processes the entities, after updating the db, rule evaluation won't be triggered again until we receive events of certain types. If we are using `minder-server` cli to update the DB, we can trigger the reevaluation for **all entities** in the DB. But if one is updating using native SQL, then that won't be possible. wdyt and what do you suggest?

Vyom-Yadav commented on proto/minder/v1/minder.proto:
```diff
@@ -863,9 +864,16 @@ message RuleEvaluationStatus {
 }
 
 message GetProfileStatusByNameRequest {
+    reserved 5;
```
I can update that, but any particular reason? I reserved them in the following format:
```
reserved 5, 6;
reserved abc, xyz;

// ... abc = 5;
// ... xyz = 6;
```
From: https://protobuf.dev/programming-guides/proto3/#deleting

> You should also reserve the field name to allow JSON and TextFormat encodings of your message to continue to parse.

Comments would also work, but it's correctness would depend on the reviewer, whereas using `reserved` keyword, the correctness would be enforced by the compiler.

evankanderson replied:
It's probably a good idea for us to reserve the names as well.  In a previous job, we only used the binary (tag number) format, so it was fine to re-use names (and sometimes desirable, as a "good name" got used for something earlier).  But I know that e.g. @AGMETEOR is using the JSON API, so we probably shouldn't change the meaning of the string names, either.

Vyom-Yadav replied:
I agree that this would break things for JSON APIS. But `RuleEvaluationStatus` has a field named `rule_name`, which is the `rule_type`. And with the introduction of `rule_name`, returning the `rule_type` in the `rule_name` field can cause a lot of unintentional errors. So, renaming would break compatibility, but it's necessary in this case. 

We can add this to release notes to make sure folks know about this change. Ideally, introducing a schema change should bump the version as well, and to serve both versions, we should have some conversion logic (taking inspiration from CRDs).

evankanderson replied:
Yeah -- we haven't _yet_ had a big enough change that we've needed to introduce API versioning.  With proto (at least) API versioning can be a bit of a pain - a bunch of copying fields that isn't clearly automateable, though Go interfaces might help over C++ or Java.

evankanderson commented:
> > However, the 2nd dependabot_configured rule is not in my rule_evaluations.
> 
> `rule_evaluations` are created and upserted by the engine. The problem is that rules for entities are evaluated again only when we receive certain events, like I mentioned in the desc:
> 
> > The rules are reconciled by the subsequent `InternalProfileInitEventTopic`, and `ExecuteEntityEventTopic` events. Profiles/Repositories not receiving these types of events **won't be reconciled** to update them to the latest state.

This is unfortunate -- it would be nice if we could trigger a re-reconcile of the rules when we do this "update".  (Ideally, we would either clear out the old evaluations that don't have a name and write new ones that did have a name.)

> > Is it by design that we only keep the one previously saved rule? Or should we create the second rule, even though there was no entry for it in rule_evaluations previously?
> 
> It is by our event-driven architecture. But I have mixed feelings about it. If we reconcile continuously and not based on events (internal, webhooks, etc.), we will consume a lot of resources as the reconciliation would continuously evaluate rules for entities. On the other hand, if we don't, then situations like these arise when the DB is not in the correct state after an update.

Edge-triggering is an optimization, but we should be aiming for a level-based (continuous reconciliation) system.  So your unease is correct.

In Kubernetes, efficiency + level-based behavior is handled by having two different mechanisms for an event to be reconciled (this is in the controller / client libraries):

1. Edge-triggered reconciliation when a resource changes (as notified by Watch in the K8s case, or by webhook in the GitHub case)
1. Background reconciliations which happen on a periodic timer, whether or not there was any activity on the resource.

We're currently missing the second, but we should be thinking about adding it, probably as some sort of background scan of all the existing rules at an infrequent (e.g. flag controlled, default to 24 hour) visitation.  We'd probably want some sort of pacing on this background scan as well -- we don't want to go into overdrive at 3AM, blow through GitHub API quota limits, and then do nothing from 3:30AM to 3AM the next day.

> The bottom line is with the current architecture and the way the engine processes the entities, after updating the db, rule evaluation won't be triggered again until we receive events of certain types. If we are using `minder-server` cli to update the DB, we can trigger the reevaluation for **all entities** in the DB. But if one is updating using native SQL, then that won't be possible. wdyt and what do you suggest?

This may be a bias in my background, but I'm used to fairly "dumb" data stores and smarter clients.  What would this look like if we did the promotion piecemeal outside the database:

1. `handlers_profile.go` will automatically create rule names on new Profiles based on the rule_type name if they are not already specified (I think you've got this).
2. `handlers_profile.go` will _also_ run the same algorithm on _output_ of existing rules to make rule names in a deterministic way.  We'd still want to flag collisions, but we might automaticallly name the rules as e.g. `<rule_type>`, `<rule_type>_1`, `<rule_type>_2`, etc.
3. Update `handlers_profile.go` to search for `rule_name = sqlc.narg(rule_name) OR rule_name IS NULL` during the migration.
4. Update `executor.go` to fill in the `rule.Name` field on rules before passing them to `TraverseRules`.  This could optionally store back to the database.

I think the interesting question is how to backfill the existing rule statuses -- two options are to simply leave the compatibility code in forever, and to backfill the database after the migration.

evankanderson commented:
BTW -- I'm looking to _discuss_ approaches in this review, not dictate them.

evankanderson commented:
With that said, it also occurs to me that the database approach here may cause a short window of breakage between when the migration script runs and when the Minder server is updated with the new queries, because the existing queries won't be including the `rule_name` field, which is `NOT NULL`.  If it somehow didn't cause breakage, that would be even more concerning, because it would mean that some profiles _could_ slip in without the `rule.name` normalization applied.

evankanderson commented:
Alternatively, since the `000013` migration is in a `BEGIN; ... COMMIT` transaction block, the other option is that this might block upgrades if there's a conflicting transaction before migration commits.

Vyom-Yadav commented:
> Background reconciliations which happen on a periodic timer, whether or not there was any activity on the resource.

Honestly, background reconciliation was the next issue I was going to pick, as we discussed that in https://github.com/stacklok/minder/pull/2109#discussion_r1449150165 too.

> 1. `handlers_profile.go` will automatically create rule names on new Profiles based on the rule_type name if they are not already specified (I think you've got this).
> 2. `handlers_profile.go` will _also_ run the same algorithm on _output_ of existing rules to make rule names in a deterministic way.  We'd still want to flag collisions, but we might automaticallly name the rules as e.g. `<rule_type>`, `<rule_type>_1`, `<rule_type>_2`, etc.
> 3. Update `handlers_profile.go` to search for `rule_name = sqlc.narg(rule_name) OR rule_name IS NULL` during the migration.
> 4. Update `executor.go` to fill in the `rule.Name` field on rules before passing them to `TraverseRules`.  This could optionally store back to the database.

Interesting, you are suggesting bringing the database to the correct state using Go. I used SQL as I didn't want to add the code of adding `name` to rows in `handlers_profile`, as it is a one-time operation while upgrading and doesn't make a lot of sense to store in a handler.

Ideally, we should have background reconciliation to solve this issue. 

> With that said, it also occurs to me that the database approach here may cause a short window of breakage between when the migration script runs and when the Minder server is updated with the new queries, because the existing queries won't be including the rule_name field, which is NOT NULL

Can we have some downtime to update the server? I don't see a way of making this update an atomic one. We could:

1. Update the server first, this would give a few errors (no crashes hopefully) when trying to `delete` or `upsert` rule evaluations as `rule_name` column doesn't exist. This would also prevent users from inserting new profiles with duplicate rules. They would be able to create profiles, but the server would throw errors while evaluating as no `rule_name` columns exist. 
2. Update the database next.

This approach ensures no bad data (without name) sneaks in while performing updates. wdyt?

Vyom-Yadav commented:
> Update the server first, this would give a few errors (no crashes hopefully) when trying to delete or upsert rule evaluations as rule_name column doesn't exist. This would also prevent users from inserting new profiles with duplicate rules. They would be able to create profiles, but the server would throw errors while evaluating as no rule_name columns exist.

This would require some modifications to the SQL script, particularly not to add `name` to `contextual rules` when it already exists.

Vyom-Yadav commented:
Also, I frequently use bold and all caps, please don't take it as shouting, it is just a way for me to highlight stuff.

evankanderson commented:
> > Background reconciliations which happen on a periodic timer, whether or not there was any activity on the resource.
> 
> Honestly, background reconciliation was the next issue I was going to pick, as we discussed that in [#2109 (comment)](https://github.com/stacklok/minder/pull/2109#discussion_r1449150165) too.

Yeah, background reconciliation has been on the "to-do" list for a while, but there are a few other features we've been trading off against (like the ability to allow multiple collaborators on the same project) and it hasn't made it to the top yet.  A PR for that would be be _awesome_.

> > 1. `handlers_profile.go` will automatically create rule names on new Profiles based on the rule_type name if they are not already specified (I think you've got this).
> > 2. `handlers_profile.go` will _also_ run the same algorithm on _output_ of existing rules to make rule names in a deterministic way.  We'd still want to flag collisions, but we might automaticallly name the rules as e.g. `<rule_type>`, `<rule_type>_1`, `<rule_type>_2`, etc.
> > 3. Update `handlers_profile.go` to search for `rule_name = sqlc.narg(rule_name) OR rule_name IS NULL` during the migration.
> > 4. Update `executor.go` to fill in the `rule.Name` field on rules before passing them to `TraverseRules`.  This could optionally store back to the database.
> 
> Interesting, you are suggesting bringing the database to the correct state using Go. I used SQL as I didn't want to add the code of adding `name` to rows in `handlers_profile`, as it is a one-time operation while updating and doesn't make a lot of sense to store in a handler.
>
> Ideally, we should have background reconciliation to solve this issue.

Yeah, that would be my preference.  I'm also coming at this more from an availability / 
 
> > With that said, it also occurs to me that the database approach here may cause a short window of breakage between when the migration script runs and when the Minder server is updated with the new queries, because the existing queries won't be including the rule_name field, which is NOT NULL
> 
> Can we have some downtime to update the server? I don't see a way of making this update an atomic one. We could:

I agree that the update can't be atomic across all database rows _and_ the server.

BTW, behind the scenes, we're running the helm chart through ArgoCD, which means that it will run the following operations:

* Run the migration script
* If the migration script succeeds, update the server replicas (multiple!)

It would be a bit tricky to actually take the server down, because ArgoCD will notice that something that's supposed to be running isn't, and will try to bring it back up.

One oddity here is that we've got an additional database we're triangulating against in terms of the GitHub API.  Since the writes to `rule_evaluations` happen _after_ we've done any work against GitHub (IIRC), we might end up doing work but not recording it if the database blocks writes (either because the `NOT NULL` column constraint, or because we're trying to write a non-existent column).

(All this operational stuff isn't normally visible to contributors, but database migrations with mandatory new columns are trickier for operational rollout than other types of PRs.)

> 1. Update the server first, this would give a few errors (no crashes hopefully) when trying to `delete` or `upsert` rule evaluations as `rule_name` column doesn't exist. This would also prevent users from inserting new profiles with duplicate rules. They would be able to create profiles, but the server would throw errors while evaluating as no `rule_name` columns exist.
> 2. Update the database next.
> 
> This approach ensures no bad data (without name) sneaks in while performing updates. wdyt?

To do this, we'd need to update the server code without the migration (adding the extra columns to the schema somehow without the migration) and add the code to write the values for new rows.  We'd then have a _second_ commit (second helm chart release) which did the migration.

BTW, [by default Postgres considers `NULL` values distinct from each other](https://www.postgresql.org/docs/current/indexes-unique.html), so you could actually do:

```sql
ALTER TABLE rule_evaluations
  ADD COLUMN rule_name TEXT;

CREATE UNIQUE INDEX rule_evaluations_results_idx2
  ON rule_evaluations (profile_id, repository_id, COALESCE(artifact_id, '00000000-0000-0000-0000-000000000000'::UUID),
                       entity, rule_type_id, COALESCE(pull_request_id, '00000000-0000-0000-0000-000000000000'::UUID),
                       rule_name);

-- Drop the existing unique index on rule_evaluations
DROP INDEX IF EXISTS rule_evaluations_results_idx;
```

in the first migration, and existing code would continue to work.  You could then do step 2 in a separate PR to backfill the existing rules.

Operationally, we'd need to note (for ourselves, and anyone else running Minder as a service) that we'd need to push between commit A and commit B, but that is do-able.  Commit B could then do the backfill and mark `rule_evaluations` `NOT NULL`.

evankanderson commented on database/migrations/000013_rule_evaluations_rule_name.up.sql:
```diff
@@ -0,0 +1,111 @@
+CREATE UNIQUE INDEX rule_evaluations_results_idx
+  ON rule_evaluations (profile_id, repository_id, COALESCE(artifact_id, '00000000-0000-0000-0000-000000000000'::UUID),
+                       entity, rule_type_id, COALESCE(pull_request_id, '00000000-0000-0000-0000-000000000000'::UUID),
+                       rule_name);
```
(We can't do that until after the backfill, though)

Not-Vyom replied:
Why do we need to care about nulls after the backfill? Nullable columns use `COALESCE` and other columns are non null.

Vyom-Yadav replied:
Why do we care about nulls after backfill? Nullable columns use `COALESCE`, and other columns are non-null.

evankanderson commented:
I'm happy to review when you're ready -- does doing this as a two-part change (add column, start filling as phase 1, then use SQL to migrate all existing rules) make sense?

Vyom-Yadav commented:
> I'm happy to review when you're ready -- does doing this as a two-part change (add column, start filling as phase 1, then use SQL to migrate all existing rules) make sense?

Yeah, the initial update would support null and non-null `rule-name`. All new profiles added would have `rule-name` populated, and the older ones would be backfilled in the next update. There were some changes required for this in the handler, would push those after testing.

Vyom-Yadav commented:
@evankanderson @eleftherias Just pinging in case you missed the review request. This and #2206 are ready for review. To give a brief summary:

1. The update will be done in two parts.
2. This PR adds support for `rule_name`. Existing rows would have `rule_name` as `null`, and newer profiles created after this update would have populated `rule_name`. The code can handle both null and non-null entires.
3. #2206 Fills the existing data. `rule_name` is calculated for profiles without one, and the evaluations are updated as well. We use an auxiliary table to support `migrate-down` of updated data.

Still, the existing data is not reconciled and is totally dependent on events to get updated to the latest state. This problem would be solved with the help of level-based reconciliation in another PR.

evankanderson reviewed (commented):
This is looking pretty good!  The two comments I care about are:

1. I repeated this a bunch of times, but doing the defaulting _before_ validation will probably simplify your logic, as right now you're trying to take into account what the defaults will be before they're written into the input.
2. I think we probably need to keep the badly-named `rule` fields in the proto for at least a month or so to give people a chance to migrate.  We can put a big comment on them that the field is going away, but we probably don't want to break compatibility for all our existing users.

The second item is kinda tricky to think about given that we have client & server in the same codebase, so it's easy to make changes and think about them atomically.  Sorry about that!

Some color that you can ignore if you want:

I've pointed out a few places where the naming is "java-flavored" (or "C#-flavored").  Go tends to towards more terse names (see the end of [package names in Effective Go](https://go.dev/doc/effective_go#package-names) for an example).  As another example, `ruleTypeAndNamePair` could probably be called `ruleKey` at no loss of understanding.  For folks using more than one programming language, it can be tricky to keep track of the different "accents", so I'm not asking for fixes here.  😁

Another style note for future PRs (don't apply it here) is to separate refactoring changes (which are super-fast to review) from logic changes (which usually require a bit more thought).  The rename of `dbstat` --> `dbProfileStatus` is an example of this:  it adds a bunch of diffs to the UI that don't really affect correctness, but make it harder to see where other code is changing (because there's already a change on the line).

Thank you again for the nice tests in `handlers_profile_test.go`!

Vyom-Yadav commented on internal/controlplane/handlers_profile.go:
```diff
@@ -96,6 +102,9 @@ func (s *Server) CreateProfile(ctx context.Context,
 		return nil, status.Errorf(codes.Internal, "error creating profile")
 	}
 
+	// Adds default rule names, if not present
+	populateRuleNames(in)
```
It would make the logic simple but, the errors won't be as clear, consider this profile:
```yaml
repository:
  - type: dependabot_configured
    def:
      package_ecosystem: gomod
  - type: dependabot_configured
    def:
      package_ecosystem: npm
```
The user has not specified the name, but as we populate before validation, we would view the profile as:
```yaml
repository:
  - type: dependabot_configured
    name: dependabot_configured
    def:
      package_ecosystem: gomod
  - type: dependabot_configured
    name: dependabot_configured
    def:
      package_ecosystem: npm
```
And we would return some error like: `rule 'dependabot_configured' has the same name 'dependabot_configured' in entity 'repository'`, which the user might get confused by as they didn't specify any name.

Now, if you say, if `ruleName == ruleType` consider user supplied empty name. But that won't be true. Users can explicitly define the name to be equal to rule type, something like:
```yaml
repository:
  - type: dependabot_configured
    name: dependabot_configured
    def:
      package_ecosystem: gomod
  - type: dependabot_configured
    def:
      package_ecosystem: npm
```
This may sound very improbable, but folks actually do stuff like this while writing any yaml. 

---

For the original profile, an error message like `cannot have rules with empty name and same type in entity 'repository'` would be better and clearer to the user. Wdyt?

evankanderson replied:
I'm okay with either error message -- we _are_ changing the schema out from under users, so a message like

```
rule 'dependabot_configured' conflicts with 'dependabot_configured' in the same profile, assign a unique name to at least one.
```

would probably not confuse users even if they hadn't been using the `name` field previously.

Vyom-Yadav replied:
Done, I modified the wording of errors to give some sort of solution as well.

Vyom-Yadav commented on internal/controlplane/handlers_profile.go:
```diff
@@ -212,7 +221,9 @@ func createProfileRulesForEntity(
 			RuleTypeID:      ruleID,
 		})
-		if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			log.Printf("the rule instantiation for rule already existed.")
```
Exactly, `entity_profile_rules` is only used to query the `rule_types` an entity in a profile uses. Elsewhere, we unmarshal `entity_profiles.contextual_rules` before processing rules.

evankanderson replied:
(I'm using these comments to help myself and future folks follow the flow of the review -- sometime in ~June, someone will do a `git blame` and come back to this PR and try to understand the decisions we made.  All of us will have forgotten, but GitHub won't!)

Vyom-Yadav commented on database/migrations/000013_rule_evaluations_rule_name.up.sql:
```diff
@@ -0,0 +1,32 @@
+-- Drop the existing unique index on rule_evaluations
+DROP INDEX IF EXISTS rule_evaluations_results_idx;
+
+-- Recreate the unique index with rule_name
+CREATE UNIQUE INDEX rule_evaluations_results_idx
+  ON rule_evaluations (profile_id, repository_id, COALESCE(artifact_id, '00000000-0000-0000-0000-000000000000'::UUID),
+                       entity, rule_type_id, COALESCE(pull_request_id, '00000000-0000-0000-0000-000000000000'::UUID),
+                       rule_name) NULLS NOT DISTINCT;
```
Correct. The reason we use `NULLS NOT DISTINCT` is because _if we don't_, two rule evaluations, which are identical would now be unique and we would have something like (omitted some cols):

| entity | profile_id | rule_type_id | rule_name | 
|--------|------------|--------------|-----------|
|   repo     |     1       |     1         |   null        |   
|   repo     |     1       |     1         |   null        |   

Now this is good for the meantime, we would also fix evaluations for existing profiles, but when we try to backfill, we can end up with:

| entity | profile_id | rule_type_id | rule_name | 
|--------|------------|--------------|-----------|
|   repo     |     1       |     1         |   abc_rule        |   
|   repo     |     1       |     1         |   abc_rule        |   

Leading to a violation of our index. That's why we require `NULLS NOT DISTINCT` .

> And since we're in a transaction, dropping the old index before adding the new is also safe; the table should only be locked briefly at our current data size.

Yes, it is safe.

Vyom-Yadav replied:
To be double sure, I checked the docs again, and:

From https://www.postgresql.org/docs/15/sql-dropindex.html
> A normal DROP INDEX acquires an ACCESS EXCLUSIVE lock on the table, blocking other accesses until the index drop can be completed. 

From https://www.postgresql.org/docs/15/explicit-locking.html
> Once acquired, a lock is normally held until the end of the transaction.

We don't have any savepoints so lock won't be released until the end of the transaction.

Vyom-Yadav commented on internal/controlplane/handlers_profile.go:
```diff
@@ -612,7 +654,7 @@ func (s *Server) UpdateProfile(ctx context.Context,
 	}
 
-	oldRules, err := s.getRulesFromProfile(ctx, oldProfile, entityCtx)
+	oldRules, err := s.getRulesFromOldProfile(ctx, oldProfile, entityCtx)
```
`getRulesFromProfile` is a generic method to get rules from a profile. It nowhere mentions whether the profile has been populated with default rule names or not. In case, it is not, then the method currently doesn't use `computeRuleName(..)` , so the returned rules won't `name` populated, which isn't ideal. Also, for old profiles, we don't want to `computeRuleName(..)`, hence the name change.

Also, this method is renamed back to `getRulesFromProfile` in the backfilling PR, when we are sure every retrieved profile from db will have a name.

Vyom-Yadav commented on internal/controlplane/handlers_profile.go:
```diff
@@ -964,3 +1071,76 @@ func deleteRuleStatusesForProfile(
+		ruleName := rule.GetName()
+		ruleType := rule.GetType()
+
+		if ruleName == ruleType {
+			return &engine.RuleValidationError{
+				Err:      fmt.Sprintf("cannot have rules with same name and type in entity '%s'", entity.ToString()),
+				RuleType: ruleType,
+			}
+		}
```
This is there to invalidate following types of profiles:
```yaml
repository:
  - type: dependabot_configured
    name: dependabot_configured
    def:
      package_ecosystem: gomod
  - type: dependabot_configured
    def:
      package_ecosystem: npm
```
And also as a general rule to make sure users have proper names in profiles.

evankanderson replied:
You're also invalidating the following profile, though:

```
repository:
  - type: dependabot_configured
    name: dependabot_configured
    def:
      package_ecosystem: gomod
```

(Which is what we'll be storing after this PR.  But actually writing the `name` field will get you an error...)

Vyom-Yadav replied:
This was intentional to prevent non required fields on the users' side. But yeah, this should not be an error, user can be verbose if they want to be.

jcchavezs commented on proto/minder/v1/minder.proto:
```diff
@@ -839,12 +839,13 @@ message ProfileStatus {
     string profile_id = 1;
     // rule_id is the id of the rule
     string rule_id = 2;
-    // rule_name is the name of the rule
-    string rule_name = 3;
```
In general this shouldn't be removed buy instead marked as deprecated.

Vyom-Yadav replied:
Marked that field as deprecated for backward compatibility.

Vyom-Yadav commented on internal/engine/eval_status.go:
```diff
@@ -73,17 +73,28 @@ func (e *Executor) createEvalStatusParams(
+	// TODO: Remove this after migration, ruleName would be valid after updating existing evaluations (#1609)
+	if len(rule.Name) == 0 {
+		ruleName.Valid = false
+	}
```
Done.

Vyom-Yadav commented on internal/controlplane/handlers_profile.go:
```diff
@@ -430,85 +442,112 @@ func (s *Server) GetProfileStatusByName(ctx context.Context,
-	if len(in.GetRule()) > 0 {
-		rule = &sql.NullString{String: in.GetRule(), Valid: true}
+	if len(in.GetRuleType()) > 0 {
+		ruleType = &sql.NullString{String: in.GetRuleType(), Valid: true}
```
Marked that field as deprecated for backward compatibility.

Vyom-Yadav commented on proto/minder/v1/minder.proto:
```diff
@@ -883,7 +891,8 @@ message GetProfileStatusByNameRequest {
     bool all = 4;
-    string rule = 5;
+    string rule_type = 6;
+    string rule_name = 7;
```
Marked that field as deprecated for backward compatibility.

Vyom-Yadav commented on internal/controlplane/handlers_profile_test.go:
```diff
@@ -0,0 +1,305 @@
+		{
+			name: "Unused old rules statuses with empty name",
+			newRules: map[ruleTypeAndNamePair]entityAndRuleTuple{
+				{RuleType: "Type1", RuleName: "Name1"}: {Entity: minderv1.Entity_ENTITY_REPOSITORIES, RuleID: generateConsistentUUID(t, "Type1", "Name1")},
+				{RuleType: "Type1", RuleName: "Name2"}: {Entity: minderv1.Entity_ENTITY_BUILD_ENVIRONMENTS, RuleID: generateConsistentUUID(t, "Type1", "Name2")},
```
There is actually a bug in the profile handler that we can't handle rules with the _same type_ in _different entities_.  Me and @JAORMX  discussed this in discord https://discord.com/channels/1184987096302239844/1185287949240242258/1199645373791219723 (Issue to be created)

evankanderson replied:
I'm not sure "rules of the same type in different entities" makes sense (because a rule_type has an `in_entity` and rule type names are unique, IIRC).  On the other hand, I'd _also_ believe that we don't validate that correctly at the moment if you told me that.

Vyom-Yadav replied:
I'll create an issue for this. We'll look at it separately

evankanderson replied:
I'd rather defer this until we have a good enough testing framework that we can write a test to show what our behavior is, and then we can feel how we feel about what the test says.

Vyom-Yadav commented on internal/controlplane/handlers_profile_test.go:
```diff
@@ -0,0 +1,305 @@
+			wantUnused: nil,
+		},
+		{
+			name: "Unused rules with same rule type",
+			newRules: map[ruleTypeAndNamePair]entityAndRuleTuple{
+				{RuleType: "Type1", RuleName: "Name1"}: {Entity: minderv1.Entity_ENTITY_REPOSITORIES, RuleID: generateConsistentUUID(t, "Type1", "Name1")},
+				{RuleType: "Type1", RuleName: "Name2"}: {Entity: minderv1.Entity_ENTITY_BUILD_ENVIRONMENTS, RuleID: generateConsistentUUID(t, "Type1", "Name2")},
+			},
+			oldRules: map[ruleTypeAndNamePair]entityAndRuleTuple{
+				{RuleType: "Type1", RuleName: "Name1"}: {Entity: minderv1.Entity_ENTITY_REPOSITORIES, RuleID: generateConsistentUUID(t, "Type1", "Name1")},
+				{RuleType: "Type1", RuleName: "Name3"}: {Entity: minderv1.Entity_ENTITY_ARTIFACTS, RuleID: generateConsistentUUID(t, "Type1", "Name3")},
+			},
+			// All rule types are used
+			wantUnused: nil,
```
Yes, it would, I added that case too.

Vyom-Yadav commented on internal/controlplane/handlers_profile_test.go:
```diff
@@ -0,0 +1,305 @@
+			},
+			wantErr:    true,
+			wantErrMsg: "cannot have rules with same name and type in entity 'repository'",
+		},
```
Added this test case.

Vyom-Yadav commented on internal/controlplane/handlers_profile_test.go:
```diff
@@ -0,0 +1,305 @@
+		entity     minderv1.Entity
+		rules      []*minderv1.Profile_Rule
+		wantErr    bool
+		wantErrMsg string
```
Done.

Vyom-Yadav commented on internal/controlplane/handlers_profile.go:
```diff
@@ -948,14 +1044,25 @@ func deleteUnusedRulesFromProfile(
+		}
+
+		// TODO: Remove this after migration, ruleName would be valid after updating existing evaluations (#1609)
+		if len(ruleTypeAndName.RuleName) == 0 {
```
Done, thanks for the information.

Vyom-Yadav commented on internal/controlplane/handlers_profile.go:
```diff
@@ -964,3 +1071,76 @@ func deleteRuleStatusesForProfile(
+				return &engine.RuleValidationError{
+					Err:      fmt.Sprintf("cannot have multiple rules in entity '%s' with same name '%s'", entity.ToString(), ruleName),
+					RuleType: ruleType,
+				}
```
Good case. Even though it won't create any problem for the backend as unique constraint won't be violated. I added logic for this and a test case as well.

Vyom-Yadav commented on internal/controlplane/handlers_profile.go:
```diff
@@ -964,3 +1071,76 @@ func deleteRuleStatusesForProfile(
+			ruleNameToType[ruleName] = ruleType
+		}
+	}
+	return nil
```
I went with validating before populating approach. I refactored the code to be more readable. This allows us to give better violation messages. Added explanatory comments too.

Vyom-Yadav commented on internal/controlplane/handlers_profile.go:
```diff
@@ -603,6 +642,9 @@ func (s *Server) UpdateProfile(ctx context.Context,
 		return nil, status.Errorf(codes.Internal, "error updating profile")
 	}
 
+	// Adds default rule names, if not present
+	populateRuleNames(in)
+
```
See https://github.com/stacklok/minder/pull/2161#discussion_r1468868810. Validating before populating allows us to give better violation messages.

evankanderson commented on internal/controlplane/handlers_profile.go:
```diff
@@ -430,85 +442,120 @@ func (s *Server) GetProfileStatusByName(ctx context.Context,
+		//nolint:staticcheck // ignore SA1019: Deprecated field supported for backward compatibility
+		ruleType = &sql.NullString{
+			String: in.GetRule(),
+			Valid:  in.GetRule() != "",
```
Non-blocking, but one of `rule_type` or `rule` should be set, right?  (Otherwise, this is very neat logic, very nice!)

Vyom-Yadav replied:
No, `rule` or `rule_type` and `rule_name` both are optional. They are associated with `--rule/--ruleType` and `--ruleName` flags respectively, so if user doesn't specify one, we return the complete eval status for a policy.

evankanderson commented on internal/engine/actions/alert/security_advisory/security_advisory.go:
```diff
@@ -86,18 +86,42 @@ To address this security exposure, we recommend taking the following actions:
+
+- Profile: {{.Profile}}
+- Rule: {{.Rule}}
+- Name: {{.Name}}
```
Sorry, I was unclear here -- I meant "don't add `tmplPart3Bottom{No,With}Name`, but don't remove the `description{NoRem}Tmpl` split in this PR.

Vyom-Yadav replied:
Done.

Not-Vyom replied:
Ah, we are using html/template, so `{{- ... -}}` won't work. Fixing it.....

Vyom-Yadav replied:
Ah, we are using html/template, so {{- ... -}} won't work. Fixing it.....

Vyom-Yadav replied:
Done, fixed.

Vyom-Yadav replied:
> In these trim markers, the white space must be present: "{{- 3}}" is like "{{3}}" but trims the immediately preceding text, while "{{-3}}" parses as an action containing the number -3.

Whitespace got me. Done, fixed.

evankanderson replied:
This now being a 7-line diff in the file fills my heart with joy, thank you!

evankanderson reviewed (approved):
I'm happy to merge this as-is unless you want to address any of the comments I made this (morning for me, evening for you)

evankanderson commented:
You'll need to merge and re-generate `minder.pb.go` with `make gen`, it looks like.

Vyom-Yadav commented:
> You'll need to merge and re-generate minder.pb.go with make gen, it looks like.

Yupp, pushing the changes...

Vyom-Yadav commented:
@evankanderson Ready for review. Changes from the last review are:

1. Modified validation logic to allow usage of `rule_type` as `name` when it does not conflict with the default name.
2. Added this point to the documentation.
3. Added additional tests for the same.
4. Used conditions in the HTML template.

evankanderson commented on docs/docs/how-to/create_profile.md:
````diff
@@ -250,4 +252,106 @@ filter the output using `jq`. For example, to get all rules that pertain to the
+If you're using multiple rules of the same type under an entity, each rule must have a unique name. This helps 
+distinguish between rules and understand their specific purpose.
+
+```yaml
+repository:
+  - type: secret_scanning
+    name: "secret_scanning_github"
+    def:
+      enabled: true
+  - type: secret_scanning
+    name: "secret_scanning_github_2"
+    def:
+      enabled: false
+```
+Here, we have two rules of the same type `secret_scanning` under the `repository` entity. Each rule has a unique name.
````
The example from https://github.com/stacklok/minder-rules-and-profiles/blob/944f5f6cf60362250e82e874e4d7c296653f2ef9/profiles/github/profile.yaml#L38 might be more relevant:

````suggestion
```yaml
repository:
  - type: dependabot_configured
    name: go_dependabot
    def:
      package_ecosystem: gomod
      schedule_interval: weekly
      apply_if_file: go.mod
  - type: dependabot_configured
    name: node_dependabot
    def:
      package_ecosystem: npm
      schedule_interval: weekly
      apply_if_file: package.json
```
Here, we have two rules of the same type `dependabot_configured` under the `repository` entity. Each rule has a unique name.
````

evankanderson reviewed (approved):
I have some tiny comments, but I'd rather get this in than fix them in this PR.

evankanderson commented on docs/docs/how-to/create_profile.md:
````diff
@@ -250,4 +252,106 @@ filter the output using `jq`. For example, to get all rules that pertain to the
+distinguish between rules and understand their specific purpose.
+
+```yaml
+repository:
+  - type: secret_scanning
+    name: "secret_scanning_github"
+    def:
+      enabled: true
+  - type: secret_scanning
+    name: "secret_scanning_github_2"
+    def:
+      enabled: false
````
Oops, I did this comment twice.  Sorry!
